src/test/run-pass/generator/borrow-in-yield-expr.rs

```diff
@@ -0,0 +1,21 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(generators, generator_trait, conservative_impl_trait)]
+
+use std::ops::Generator;
+
+fn bar(baz: String) -> impl Generator<Yield=(), Return=()> {
+    move || {
+        yield drop(&baz);
```

Comment:
I'm afraid you can do
```Rust
struct DropBomb;
impl Drop for DropBomb {
    fn drop(&mut self) {
        println!("splat!");
    }
}

fn bar() -> impl Generator<Yield=(), Return=()> {
    move || {
        yield &DropBomb;
    }
}
```

Because `yield` is currently not a terminating scope, this would only get dropped *after* the generator returns. If `DropBomb` was e.g. `!Send`, this can obviously cause trouble. 

This can be solved by making `yield` a terminating scope, which would force temporaries in all of its arguments to be destroyed before it is reached (search for `terminating_scope` in `region`).

Reply:
We may want to allow generators to return references to locals, in which case having  `DropBomb` is live across the suspension point would be useful. I changed the ordering so the suspension point comes after the yield subexpression making `&DropBomb` and `DropBomb` a part of the generator interior.